modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/ContributorsBuilder.es.js

```diff
@@ -0,0 +1,282 @@
+import getCN from 'classnames';
+import React from 'react';
+import PropTypes from 'prop-types';
+import {DragDropContext as dragDropContext} from 'react-dnd';
+import HTML5Backend from 'react-dnd-html5-backend';
+import CriteriaSidebar from '../criteria_sidebar/CriteriaSidebar.es';
+import {sub} from '../../utils/utils.es';
+import CriteriaBuilder from './CriteriaBuilder.es';
+import {buildQueryString, translateQueryToCriteria} from '../../utils/odata.es';
+import Conjunction from './Conjunction.es';
+
+class ContributorsBuilderComp extends React.Component {
+	constructor(props) {
+		super(props);
+		this.classes = getCN(
+			'criteria-builder-root',
+			{
+				'read-only': false
+			}
+		);
+		const {initialContributors, propertyGroups} = props;
+		const contributors = initialContributors && initialContributors.map(
+			c => {
+				const propertyGroup = propertyGroups && propertyGroups.find(t => c.propertyKey === t.propertyKey);
+
+				return {
+					conjunctionId: c.conjunctionId || 'and',
+					conjunctionInputId: c.conjunctionInputId,
+					criteriaMap: c.initialQuery ?
+						translateQueryToCriteria(c.initialQuery) :
+						null,
+					inputId: c.inputId,
+					modelLabel: propertyGroup && propertyGroup.name,
+					properties: propertyGroup && propertyGroup.properties,
+					propertyKey: c.propertyKey,
+					query: c.initialQuery
+				};
+			}
+		);
+		this.onCriteriaEdit = this.onCriteriaEdit.bind(this);
+		this.onCriteriaChange = this.onCriteriaChange.bind(this);
+		this.state = {
+			conjunctionName: 'and',
+			contributors,
+			editing: undefined,
+			newPropertyKey: this.props.propertyGroups.length &&
+				this.props.propertyGroups[0].propertyKey
+		};
+		this._handleRootConjunctionClick = this._handleRootConjunctionClick.bind(this);
+		this._handleSelectorChange = this._handleSelectorChange.bind(this);
+	}
+
+	onCriteriaEdit(id, editing) {
+		this.setState(
+			{
+				editing: editing ? undefined : id
+			}
+		);
+	}
+
+	_handleSelectorChange(e) {
+		const newPropertyKey = e.target.value;
+
+		this.setState(
+			prevState => (
+				{
+					...prevState,
+					newPropertyKey
+				}
+			)
+		);
+	}
+
+	onCriteriaChange(criteriaChange, index) {
+		this.setState(
+			state => {
+				let newState;
+				if (state.editing !== index) {
+					newState = state;
```

Comment:
Rather than calling setState with no changes, we should move this `if()` outside of setState. For example

```js
onCriteriaChange(criteriaChange, index) {
  if (this.state.editing === index) {
    this.setState({
      contributors: this.state.contributors.map((c, i) => {
        return index === i
          ? {
              ...c,
              criteriaMap: criteriaChange,
              query: buildQueryString([criteriaChange])
            }
          : c
      })
    })
  }
}
```

Reply:
totally missed this comment. The `if` can go out, but I wouldn't rely on `this.state` inside `setState`.

I'm treating the state as a reducer (always returning the full state), and that's wrong, I should return only the diffs. We can avoid any unnecessary work making that `setState` returning `null` when there shouldn't be any update.